### PR TITLE
[36238] Extract and fix user references in other objects

### PR DESCRIPTION
**Pull request by oliverguenther**
- [x] Add common destroy job for all principals
- [x] That job calls the extracted reference updating that was a `before_destroy` hook for users.
- [x] Adds migration to fix the issues with dangling `custom_values` and `responsible_id` references.  

This PR also unifies the way `assigned_to` and `responsible` are being handled on deletion. For users, the references were nullified. For groups, the deleted user was assigned. Now, in any case the deleted user is going to be assigned.

https://community.openproject.com/work_packages/36238

### CODE REVIEW

File: app/workers/principals/replace_references_job.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+#++
+
+module Principals
+  class ReplaceReferencesJob < ApplicationJob
+    queue_with_priority :low
+
+    def perform(principal_id)
+      reassign_author principal_id
+      reassign_user_id principal_id
+      reassign_user_custom_values principal_id
```
Comment (wielinde) [inline app/workers/principals/replace_references_job.rb]: I believe that some `reassign_responsible` method is missing for WP responsibles

Comment (wielinde) [reply app/workers/principals/replace_references_job.rb]: Oh, you tackle that in later commit.

File: spec/models/group_spec.rb [hunk 2]
```diff
@@ -81,15 +81,7 @@
 
     it 'should roles removed when removing group membership' do
       expect(user).to be_member_of project
-      member.destroy
```
Comment (wielinde) [inline spec/models/group_spec.rb:84]: Destroying a member of a group is not the same as destroying the group. There were two different test cases before this commit. Now it is only one. Testing the deletion of a group member is gone and I believe it should be there.

File: spec/models/group_spec.rb [hunk 3]
```diff
@@ -128,15 +120,15 @@
       end
 
       it 'should reassign the work package to nobody' do
```
Comment (wielinde) [inline spec/models/group_spec.rb]: The example should be renamed from 
'should reassign the work package to nobody'
to
'should reassign the work package to DeletedUser',
I believe.

File: spec/models/group_spec.rb (hunk 2 above)
Comment (oliverguenther) [reply spec/models/group_spec.rb:84]: Well spotted, I was sure they were duplicate when I removed them

**Comment by oliverguenther** [comment]
@wielinde I addressed your remarks

**Pull request merged**